File: etl/wald.py
"""
Waldfläche Österreichs und seiner Nachbarn — Eurostat for_area (FAO).

ECHTE API. Der einzige geprüfte Biodiversitätsdatensatz, der auch die
Nicht-EU-Nachbarn Schweiz UND Liechtenstein mit Werten führt — bei den
Schutzgebieten (`sdg_15_20`) sind beide zwar als Dimension gelistet, aber
ohne Daten. Wer das übersieht, baut eine Nachbarschaftsgrafik mit zwei
stillen Lücken.

Nur sechs Stützjahre (1990, 2000, 2010, 2015, 2020, 2025), dafür über 35
Jahre. Gezeigt wird die VERÄNDERUNG seit 1990 in Prozent, nicht die
absolute Fläche: Deutschland hat dreimal so viel Wald wie Österreich, das
sagt über die Entwicklung nichts.

WAS DIE ZAHL NICHT SAGT — und der Hinweistext muss es tragen:
Waldfläche ist eine Mengenangabe, kein Zustandsmaß. Eine Fichtenmonokultur
zählt so viel wie ein Auwald. Der eigentliche Biodiversitätsindikator im
Wald ist das Totholz, und dessen Anteil am Holzvorrat liegt in Österreich
laut Waldinventur 2018–2023 bei rund 3 % — abrufbar ist diese Reihe nur
über das interaktive Werkzeug des BFW, nicht über eine Schnittstelle.
"""

from __future__ import annotations

import config
from gemeinsam import (jsonstat_laender, lade_json, laender_namen, log,
                       quelle_vermerken, warnen)

# Anteil des stehenden Totholzes am Holzvorrat, Österreichische Waldinventur
# 2018–2023. Von waldinventur.at abgelesen (24.08.2026); die Reihe selbst
# gibt das Werkzeug nicht als Datei heraus.
TOTHOLZ_ANTEIL = 3.0
TOTHOLZ_PERIODE = "2018–2023"


def baue_wald() -> dict | None:
    log("\n[7/8] Waldfläche — Österreich und Nachbarn (Eurostat for_area)")

    url = f"{config.EUROSTAT_BASIS}/{config.WALD_CODE}"
    roh = lade_json(url, config.WALD_PARAMS)
    laender = jsonstat_laender(roh, "for_area")
    namen = laender_namen(roh)

    if not laender:
        warnen("Waldfläche: keine Werte — Abschnitt bleibt ausgeblendet")
        return None

    # KEIN EU-Aggregat: `for_area` führt ausschließlich einzelne Staaten,
    # eine EU-Summe gibt es in diesem Datensatz nicht (geprüft 24.08.2026).
    # Sie mitanzufordern erzeugte bei jedem Lauf eine Warnung über eine
    # Lücke, die gar keine ist.
    gesucht = list(config.NACHBARN)
    fehlend = [code for code in gesucht if code not in laender]
    if fehlend:
        # Melden statt still weglassen: eine fehlende Zeile in der Grafik
        # sieht aus wie „kein Wald", nicht wie „keine Meldung".
        warnen(
            f"Waldfläche: keine Werte für {', '.join(fehlend)} — "
            f"diese Gebiete fehlen im Vergleich"
        )

    eintraege = []
    for code in gesucht:
        reihe = laender.get(code)
        if not reihe:
            continue
        jahre = sorted(reihe)
        erst, letzt = jahre[0], jahre[-1]
        basis, jetzt = reihe[erst], reihe[letzt]
        if not basis:
            continue
        eintraege.append({
            "code": code,
            "name": namen.get(code, code),
            "von": int(erst),
            "bis": int(letzt),
            "flaeche_von": round(basis, 1),
            "flaeche_bis": round(jetzt, 1),
            "veraenderung": round((jetzt - basis) / basis * 100, 1),
            "hervorgehoben": code == config.HERVORHEBUNG,
            "nur_ein_jahr": len(jahre) < 2,
            "reihe": [{"jahr": int(j), "wert": round(reihe[j], 1)} for j in jahre],
        })

    if not eintraege:
        warnen("Waldfläche: keiner der gesuchten Staaten hat Werte")
        return None

    eintraege.sort(key=lambda e: e["veraenderung"], reverse=True)
    oesterreich = next((e for e in eintraege if e["code"] == "AT"), None)

    if oesterreich:
        rang = eintraege.index(oesterreich) + 1
        log(f"    Österreich: {oesterreich['flaeche_von']:,.0f} → "
            f"{oesterreich['flaeche_bis']:,.0f} Tsd. ha "
            f"({oesterreich['veraenderung']:+.1f} %), Platz {rang} von "
            f"{len(eintraege)}")
    else:
        rang = None
        warnen("Waldfläche: Österreich fehlt in der Antwort")

    quelle_vermerken(
        name="Eurostat — for_area, Waldfläche (Quelle: FAO)",
        url="https://ec.europa.eu/eurostat/databrowser/view/for_area",
        lizenz="Eurostat-Nutzungsbedingungen",
        stand=str(eintraege[0]["bis"]),
        art="api",
    )

    return {
        "eintraege": eintraege,
        "oesterreich": oesterreich,
        "rang": rang,
        "anzahl": len(eintraege),
        "von": eintraege[0]["von"],
        "bis": eintraege[0]["bis"],
        "totholz_anteil": TOTHOLZ_ANTEIL,
        "totholz_periode": TOTHOLZ_PERIODE,
        "hinweis": (
            "Waldfläche ist eine Mengenangabe, kein Zustandsmaß — eine "
            "Fichtenmonokultur zählt so viel wie ein Auwald. Der eigentliche "
            f"Indikator im Wald ist das Totholz; sein Anteil am Holzvorrat "
            f"liegt in Österreich bei rund {TOTHOLZ_ANTEIL:.0f} % "
            f"({TOTHOLZ_PERIODE})."
        ),
    }

File: docs/js/charts/wald.js
/* ===========================================================================
   Biodiversitäts-Dashboard Österreich — Themenstrang: wald
   ---------------------------------------------------------------------------
   Wird nach js/kern.js geladen; die Helfer kommen aus window.BIO.
   =========================================================================== */
(function (BIO) {
"use strict";
const { stil, zahl, pz, basis, achse, tabelle, setzeText, setzeHtml,
        diagramme, schrift, balkenGitter, kategorieLabel, balkenBreite, balkenHoehe, hoverDunkler } = BIO;

/* --- 07 — Waldfläche: Österreich und seine Nachbarn ------------------
   Liegende Balken, Veränderung der Waldfläche seit 1990 in Prozent.

   WARUM VERÄNDERUNG UND NICHT FLÄCHE: Deutschland hat dreimal so viel
   Wald wie Österreich und Liechtenstein ein Tausendstel davon. Absolute
   Flächen nebeneinander sagen über die Entwicklung nichts und machen die
   kleinen Länder unsichtbar. Prozent macht sie vergleichbar.

   WARUM DIESER DATENSATZ: `for_area` ist der einzige geprüfte
   Biodiversitätsdatensatz, der auch die Nicht-EU-Nachbarn Schweiz und
   Liechtenstein mit echten Werten führt. Bei den Schutzgebieten sind beide
   zwar als Dimension gelistet, aber ohne Daten — wer das übersieht, baut
   eine Nachbarschaftsgrafik mit zwei stillen Lücken.

   ÖSTERREICH IST HERVORGEHOBEN, der Rest ist Kontext. Ein Ranking, in dem
   alle Balken gleich aussehen, zwingt zum Suchen.

   ZUR NULLLINIE: Sie ist hier bedeutungstragend — links davon schrumpft
   der Wald, rechts wächst er. Deshalb steht sie als eigene Achslinie und
   nicht nur als Rasterlinie. */

function baueWald(daten) {
  const S = schrift();
  if (!daten?.eintraege?.length) return;
  const abschnitt = document.getElementById("s-wald");
  if (abschnitt) abschnitt.style.display = "";

  const feld = document.getElementById("c-wald");
  if (!feld) return;
  const d = echarts.getInstanceByDom(feld) || echarts.init(feld, null, { renderer: "svg" });
  if (!diagramme.includes(d)) diagramme.push(d);

  const eintraege = daten.eintraege;
  const hell = stil("--viz-series-3");
  const dunkel = stil("--viz-series-1");

  setzeText("u-wald",
    `Veränderung der Waldfläche ${daten.von} bis ${daten.bis} · ` +
    `Österreich und seine acht Nachbarn`);
  setzeText("h-wald", daten.hinweis ?? "");

  /* Eng braucht jede Kategorie eine eigene Zeile fuer ihren Namen.
     Die Kartenhoehe kommt deshalb aus der Zahl der Kategorien und
     nicht aus dem CSS — sonst schiebt ECharts die Zeilen enger
     zusammen, als der Name hoch ist, und die Namen kleben am
     Balken der Zeile darueber. Muss VOR setOption stehen. */
  balkenHoehe(d, feld, eintraege.length, 2);

  d.setOption({
    ...basis(),
    grid: { ...balkenGitter(feld, { left: 128, right: 72 }), top: 12 },
    tooltip: {
      ...basis().tooltip, trigger: "axis",
      axisPointer: { type: "shadow", shadowStyle: { color: stil("--viz-grid"), opacity: 0.35 } },
      formatter: (p) => {
        const e = eintraege[p[0].dataIndex];
        if (e.nur_ein_jahr) {
          return `<strong>${e.name}</strong><br>` +
            `<span style="color:${stil("--viz-muted")}">Nur ein Stützjahr ` +
            `(${e.bis}) gemeldet — keine Veränderung berechenbar</span>`;
        }
        return `<strong>${e.name}</strong><br>` +
          `<strong>${e.veraenderung > 0 ? "+" : ""}${pz(e.veraenderung)} %</strong> ` +
          `seit ${e.von}<br>` +
          `<span style="color:${stil("--viz-muted")}">` +
          `${zahl(Math.round(e.flaeche_von))} → ${zahl(Math.round(e.flaeche_bis))} ` +
          `Tausend Hektar</span>`;
      },
    },
    xAxis: { ...achse(), type: "value",
      axisLine: { show: false },
      axisLabel: { hideOverlap: true, color: stil("--viz-muted"),
                   fontSize: S.achse,
                   formatter: (v) => (v > 0 ? "+" : "") + v + " %" } },
    yAxis: { ...achse(), type: "category", inverse: true,
      data: eintraege.map((e) => e.name), splitLine: { show: false },
      axisLabel: { fontSize: S.serie, margin: 12,
                   color: stil("--viz-text-2"),
                   ...kategorieLabel(feld, 128, eintraege.length) } },
    series: [{
      type: "bar", name: "Veränderung", barWidth: balkenBreite(feld, "58%", eintraege.length),
      data: eintraege.map((e) => ({
        value: e.veraenderung,
        itemStyle: {
          color: e.hervorgehoben ? dunkel : hell,
          borderRadius: e.veraenderung >= 0 ? [0, 4, 4, 0] : [4, 0, 0, 4],
        },
        emphasis: hoverDunkler(e.hervorgehoben ? dunkel : hell),
      })),
      label: { show: true, position: "right", distance: 8,
               color: stil("--viz-text-2"), fontSize: S.label,
               formatter: (p) => (p.value > 0 ? "+" : "") + pz(p.value) + " %" },
      /* Nulllinie sichtbar machen: Sie trennt Wachstum von Rückgang. */
      markLine: {
        silent: true, symbol: "none",
        lineStyle: { color: stil("--viz-axis"), width: 1 },
        label: { show: false },
        data: [{ xAxis: 0 }],
      },
    }],
  }, { replaceMerge: ["series", "xAxis", "yAxis"] });

  const at = daten.oesterreich;
  if (at) {
    /* 28.08.2026 — der Schlusssatz („Dass die Fläche wächst, sagt nichts
       darüber, was auf ihr wächst") ist entfallen. Die Einordnung
       darunter sagt dasselbe, nur konkret: „eine Fichtenmonokultur zählt
       so viel wie ein Auwald". Von zwei Fassungen desselben Gedankens
       bleibt die mit dem Bild. */
    setzeHtml("n-wald",
      `Österreichs Waldfläche wuchs zwischen ${at.von} und ${at.bis} um ` +
      `<strong>${pz(at.veraenderung)} %</strong> auf ${zahl(Math.round(at.flaeche_bis))} ` +
      `Tausend Hektar — Platz ${daten.rang} von ${daten.anzahl} im ` +
      `Nachbarschaftsvergleich.`);
  }

  setzeHtml("t-wald", tabelle(
    [{ titel: "Land", wert: (z) => z.name },
     { titel: `${daten.von} (Tsd. ha)`, num: true,
       wert: (z) => z.nur_ein_jahr ? "–" : zahl(Math.round(z.flaeche_von)) },
     { titel: `${daten.bis} (Tsd. ha)`, num: true,
       wert: (z) => zahl(Math.round(z.flaeche_bis)) },
     { titel: "Veränderung", num: true,
       wert: (z) => z.nur_ein_jahr ? "–"
         : (z.veraenderung > 0 ? "+" : "") + pz(z.veraenderung) + " %" }],
    eintraege
  ));
}

BIO.baueWald = baueWald;
})(window.BIO);
